gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java

```diff
@@ -146,6 +156,95 @@ protected void addLineageSourceInfo(SourceState sourceState, SourceEntity entity
 
   @Override
   protected List<WorkUnit> generateWorkUnits(SourceEntity sourceEntity, SourceState state, long previousWatermark) {
+    String partitionType = state.getProp(PARTITION_TYPE, "PK_CHUNKING");
+    if (partitionType.equals("PK_CHUNKING")) {
+      return generateWorkUnitsPkChunking(sourceEntity, state, previousWatermark);
+    } else {
+      return generateWorkUnitsStrategy(sourceEntity, state, previousWatermark);
+    }
+  }
+
+  /**
+   * generate workUnit with noQuery=true
+   */
+  private List<WorkUnit> generateWorkUnitsPkChunking(SourceEntity sourceEntity, SourceState state, long previousWatermark) {
+      List<SalesforceExtractor.BatchIdAndResultId> batchIdAndResultIds = executeQueryWithPkChunking(state, previousWatermark);
+      List<WorkUnit> ret = createWorkUnits(sourceEntity, state, batchIdAndResultIds);
+      return ret;
+  }
+
+  private List<SalesforceExtractor.BatchIdAndResultId> executeQueryWithPkChunking(
+      SourceState sourceState,
+      long previousWatermark
+  ) throws RuntimeException {
+    Properties commonProperties = sourceState.getCommonProperties();
+    Properties specProperties = sourceState.getSpecProperties();
+    State state = new State();
+    state.setProps(commonProperties, specProperties);
+    WorkUnit workUnit = WorkUnit.createEmpty();
+    try {
+      WorkUnitState workUnitState = new WorkUnitState(workUnit, state);
+      workUnitState.setId("test" + new Random().nextInt());
+      workUnitState.setProp(ENABLE_PK_CHUNKING_KEY, true); // set extractor enable pk chunking
+      int chunkSize = workUnitState.getPropAsInt(PARTITION_PK_CHUNKING_SIZE, DEFAULT_PK_CHUNKING_SIZE);
```

Comment:
Is it necessary to introduce another config key to chunk size?

Reply:
I was thinking we may want to keep 2nd level PK-chunking and better have different property for them.
As we discussed, we don't 2nd level PK-chunking doesn't make sense. Will remove this property.